python/src/scipp/compat/dict.py

```
@@ -0,0 +1,227 @@
+            out[key][str(name)] = _variable_to_dict(item)
+    if da.unaligned is not None:
+        out["unaligned"] = _data_array_to_dict(da.unaligned)
+    else:
+        out["data"] = _variable_to_dict(da.data)
+    out["name"] = da.name
+    return out
+
+
+def _dims_to_strings(dims):
+    """
+    Convert dims that may or may not be strings to strings.
+    """
+    return [str(dim) for dim in dims]
+
+
+def from_dict(dict_obj):
+    """
+    Convert a python dict to a scipp Variable, DataArray or Dataset.
+    If the input keys contain both `'coords'` and `'data'`, then a DataArray is
+    returned.
+    If the input keys contain both `'dims'` and `'values'`, as Variable is
+    returned.
+    Otherwise, a Dataset is returned.
+
+    :param dict_obj: A python dict to be converted to a scipp object.
+    :type dict_obj: dict
+    :return: A scipp Variable, DataArray or Dataset.
+    :rtype: Variable, DataArray, or Dataset
+    """
+    keys_as_set = set(dict_obj.keys())
+    if ({"coords", "data"}.issubset(keys_as_set)
+            or {"coords", "unaligned"}.issubset(keys_as_set)):
+        # Case of a DataArray-like dict (most-likely)
+        return _dict_to_data_array(dict_obj)
+    elif (keys_as_set.issubset(
+        {"dims", "values", "variances", "unit", "dtype", "shape"})
+          or keys_as_set.issubset(
+              {"value", "variance", "unit", "dtype", "shape", "dims"})):
+        # Case of a Variable-like dict (most-likely)
+        return _dict_to_variable(dict_obj)
+    else:
+        # Case of a Dataset-like dict
+        out = sc.Dataset()
+        for key, item in dict_obj.items():
+            out[key] = _dict_to_data_array(item)
+        return out
+
+
+def _dict_to_variable(d):
+    """
+    Convert a python dict to a scipp Variable.
+    """
+    # The Variable constructor does not accept both `shape` and `values`. If
+    # `values` is present, remove `shape` from the list. Also remove `dims` in
+    # the case of a 0D variable.
```

So this results in silently ignoring potentially mismatching information. I think we have two options:
- Improve the `Variable` constructor so this is allowed.
- Do not support it here.